tests/linkediting.js

```diff
@@ -104,6 +118,189 @@ describe( 'LinkEditing', () => {
+
+			view.document.fire( 'paste', {
+				dataTransfer: dataTransferMock,
+				preventDefault() {},
+				stopPropagation() {}
+			} );
+
+			expect( getModelData( model ) ).to.equal( '<paragraph><$text linkHref="ckeditor.com">foPASTED[]o</$text></paragraph>' );
+			expect( model.document.selection.isGravityOverridden ).to.be.false;
+		} );
+
+		it( 'should not override the gravity when pasting before a link when the gravity is already overridden', () => {
+			const dataTransferMock = createDataTransfer( { 'text/html': '<b>PASTED</b>' } );
+
+			setModelData( model, '<paragraph>foo[]<$text linkHref="ckeditor.com">bar</$text></paragraph>' );
+
+			view.document.fire( 'keydown', {
+				keyCode: keyCodes.arrowright,
+				preventDefault: () => {},
+				domTarget: document.body
+			} );
+
+			expect( model.document.selection.isGravityOverridden ).to.be.true;
+
+			view.document.fire( 'paste', {
+				dataTransfer: dataTransferMock,
+				preventDefault() {},
+				stopPropagation() {}
+			} );
+
+			expect( getModelData( model ) ).to.equal(
+				'<paragraph>' +
+					'foo' +
+					'<$text bold="true">PASTED</$text>' +
+					'<$text linkHref="ckeditor.com">[]bar</$text>' +
+				'</paragraph>'
+			);
+
+			expect( model.document.selection.isGravityOverridden ).to.be.true;
+		} );
+
+		it( 'should not override the gravity when pasting a link into another link (different URLs, they won\'t merge)', () => {
+			const dataTransferMock = createDataTransfer( { 'text/html': '<a href="http://PASTED">PASTED</a>' } );
+
+			setModelData( model, '<paragraph><$text linkHref="ckeditor.com">f[]oo</$text></paragraph>' );
+
+			view.document.fire( 'paste', {
+				dataTransfer: dataTransferMock,
+				preventDefault() {},
+				stopPropagation() {}
+			} );
+
+			expect( getModelData( model ) ).to.equal(
+				'<paragraph>' +
+					'<$text linkHref="ckeditor.com">f</$text>' +
+					'<$text linkHref="http://PASTED">PASTED[]</$text>' +
+					'<$text linkHref="ckeditor.com">oo</$text>' +
+				'</paragraph>'
+			);
+
+			expect( model.document.selection.isGravityOverridden ).to.be.false;
+		} );
+
+		it( 'should not override the gravity when pasting before another link (different URLs, they won\'t merge)', () => {
+			const dataTransferMock = createDataTransfer( { 'text/html': '<a href="http://PASTED">PASTED</a>' } );
+
+			setModelData( model, '<paragraph>[]<$text linkHref="ckeditor.com">foo</$text></paragraph>' );
+
+			expect( model.document.selection.isGravityOverridden ).to.be.false;
+
+			view.document.fire( 'paste', {
+				dataTransfer: dataTransferMock,
+				preventDefault() {},
+				stopPropagation() {}
+			} );
+
+			expect( getModelData( model ) ).to.equal(
+				'<paragraph>' +
+					'<$text linkHref="http://PASTED">PASTED[]</$text>' +
+					'<$text linkHref="ckeditor.com">foo</$text>' +
+				'</paragraph>'
+			);
+
+			expect( model.document.selection.isGravityOverridden ).to.be.false;
+		} );
+
+		it( 'should restore gravity on the next change:range', () => {
```

Comment:
Looks like the test description is too technical ;) and will need to change. I'd go with the expected behaviour like when user move selection, or something similar.